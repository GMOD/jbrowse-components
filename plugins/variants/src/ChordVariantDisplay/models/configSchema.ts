import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { baseChordDisplayConfig } from '@jbrowse/plugin-circular-view'
import { types } from 'mobx-state-tree'
import type PluginManager from '@jbrowse/core/PluginManager'

/**
 * #config ChordVariantDisplay
 */
function x() {} // eslint-disable-line @typescript-eslint/no-unused-vars

function stateModelFactory(pluginManager: PluginManager) {
  return ConfigurationSchema(
    'ChordVariantDisplay',
    {
      /**
       * #slot
       */
      renderer: types.optional(
        pluginManager.pluggableConfigSchemaType('renderer'),
        { type: 'StructuralVariantChordRenderer' },
      ),
    },
    {
      /**
       * #baseConfiguration
       */
      baseConfiguration: baseChordDisplayConfig,
      explicitlyTyped: true,
    },
  )
}

export default stateModelFactory
