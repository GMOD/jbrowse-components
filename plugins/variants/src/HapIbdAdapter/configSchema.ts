import { ConfigurationSchema } from '@jbrowse/core/configuration'

/**
 * #config HapIbdAdapter
 */
function x() {} // eslint-disable-line @typescript-eslint/no-unused-vars

const HapIbdAdapter = ConfigurationSchema(
  'HapIbdAdapter',
  {
    /**
     * #slot
     */
    hapIbdLocation: {
      type: 'fileLocation',
      defaultValue: {
        uri: '/path/to/my.hap.ibd.gz',
        locationType: 'UriLocation',
      },
    },
  },
  { explicitlyTyped: true },
)

export default HapIbdAdapter
