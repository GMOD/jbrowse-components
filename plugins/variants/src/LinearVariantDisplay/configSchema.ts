import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { types, Instance } from 'mobx-state-tree'
import PluginManager from '@jbrowse/core/PluginManager'
import { baseLinearDisplayConfigSchema } from '@jbrowse/plugin-linear-genome-view'

export function LinearVariantDisplayConfigFactory(
  pluginManager: PluginManager,
) {
  const PileupRendererConfigSchema = pluginManager.getRendererType(
    'PileupRenderer',
  ).configSchema
  const SvgFeatureRendererConfigSchema = pluginManager.getRendererType(
    'SvgFeatureRenderer',
  ).configSchema

  return ConfigurationSchema(
    'LinearVariantDisplay',
    {
      defaultRendering: {
        type: 'stringEnum',
        model: types.enumeration('Rendering', ['pileup', 'svg']),
        defaultValue: 'svg',
      },

      renderers: ConfigurationSchema('RenderersConfiguration', {
        PileupRenderer: PileupRendererConfigSchema,
        SvgFeatureRenderer: SvgFeatureRendererConfigSchema,
      }),
    },
    { baseConfiguration: baseLinearDisplayConfigSchema, explicitlyTyped: true },
  )
}

export type LinearVariantDisplayConfigModel = ReturnType<
  typeof LinearVariantDisplayConfigFactory
>
export type LinearVariantDisplayConfig = Instance<
  LinearVariantDisplayConfigModel
>
