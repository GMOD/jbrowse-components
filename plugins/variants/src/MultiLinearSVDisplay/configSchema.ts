import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { types } from 'mobx-state-tree'

// locals
import sharedVariantConfigFactory from '../shared/SharedVariantConfigSchema'

import type PluginManager from '@jbrowse/core/PluginManager'

/**
 * #config MultiLinearSVDisplay
 * extends
 * - [SharedSVDisplay](../sharedvariantdisplay)
 */
function x() {} // eslint-disable-line @typescript-eslint/no-unused-vars

export default function SVConfigFactory(pluginManager: PluginManager) {
  const MultiSVRendererConfigSchema =
    pluginManager.getRendererType('MultiSVRenderer')!.configSchema

  return ConfigurationSchema(
    'MultiLinearSVDisplay',
    {
      /**
       * #slot
       */
      defaultRendering: {
        type: 'stringEnum',
        model: types.enumeration('Rendering', ['multivariant']),
        defaultValue: 'multivariant',
      },

      /**
       * #slot
       */
      renderers: ConfigurationSchema('RenderersConfiguration', {
        MultiSVRenderer: MultiSVRendererConfigSchema,
      }),

      /**
       * #slot
       */
      height: {
        type: 'number',
        defaultValue: 200,
      },
    },
    {
      /**
       * #baseConfiguration
       */
      baseConfiguration: sharedVariantConfigFactory(),
      explicitlyTyped: true,
    },
  )
}
