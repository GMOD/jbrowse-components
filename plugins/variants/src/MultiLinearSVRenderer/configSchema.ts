import { ConfigurationSchema } from '@jbrowse/core/configuration'

/**
 * #config MultiSVRenderer
 */
function x() {} // eslint-disable-line @typescript-eslint/no-unused-vars

const configSchema = ConfigurationSchema(
  'MultiSVRenderer',
  {},
  {
    /**
     * #baseConfiguration
     */
    explicitlyTyped: true,
  },
)

export default configSchema
