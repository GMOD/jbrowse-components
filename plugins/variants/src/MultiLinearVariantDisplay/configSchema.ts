import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { types } from 'mobx-state-tree'

// locals
import sharedVariantConfigFactory from '../shared/SharedVariantConfigSchema'

import type PluginManager from '@jbrowse/core/PluginManager'

/**
 * #config MultiLinearVariantDisplay
 * extends
 * - [SharedVariantDisplay](../sharedvariantdisplay)
 */
function x() {} // eslint-disable-line @typescript-eslint/no-unused-vars

export default function VariantConfigFactory(pluginManager: PluginManager) {
  const MultiVariantRendererConfigSchema = pluginManager.getRendererType(
    'MultiVariantRenderer',
  )!.configSchema

  return ConfigurationSchema(
    'MultiLinearVariantDisplay',
    {
      /**
       * #slot
       */
      defaultRendering: {
        type: 'stringEnum',
        model: types.enumeration('Rendering', ['multivariant']),
        defaultValue: 'multivariant',
      },

      /**
       * #slot
       */
      renderers: ConfigurationSchema('RenderersConfiguration', {
        MultiVariantRenderer: MultiVariantRendererConfigSchema,
      }),

      /**
       * #slot
       */
      height: {
        type: 'number',
        defaultValue: 200,
      },
    },
    {
      /**
       * #baseConfiguration
       */
      baseConfiguration: sharedVariantConfigFactory(),
      explicitlyTyped: true,
    },
  )
}
