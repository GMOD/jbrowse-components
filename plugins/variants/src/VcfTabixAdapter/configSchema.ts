import { types } from 'mobx-state-tree'
import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'VcfTabixAdapter',
  {
    vcfGzLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/my.vcf.gz' },
    },
    index: ConfigurationSchema('VcfIndex', {
      indexType: {
        model: types.enumeration('IndexType', ['TBI', 'CSI']),
        type: 'stringEnum',
        defaultValue: 'TBI',
      },
      location: {
        type: 'fileLocation',
        defaultValue: { uri: '/path/to/my.vcf.gz.tbi' },
      },
    }),
  },
  { explicitlyTyped: true },
)
