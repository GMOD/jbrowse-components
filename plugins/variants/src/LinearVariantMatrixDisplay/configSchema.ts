import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { linearBasicDisplayConfigSchemaFactory } from '@jbrowse/plugin-linear-genome-view'

// locals
import configSchema from '../LinearVariantMatrixRenderer/configSchema'

import type PluginManager from '@jbrowse/core/PluginManager'

/**
 * #config LinearVariantMatrixDisplay
 */
function x() {} // eslint-disable-line @typescript-eslint/no-unused-vars

export default function configSchemaF(pluginManager: PluginManager) {
  return ConfigurationSchema(
    'LinearVariantMatrixDisplay',
    {
      renderer: configSchema,
    },
    {
      /**
       * #baseConfiguration
       */
      baseConfiguration: linearBasicDisplayConfigSchemaFactory(pluginManager),
      explicitlyTyped: true,
    },
  )
}
