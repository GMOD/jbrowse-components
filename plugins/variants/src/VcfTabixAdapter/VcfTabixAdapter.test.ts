import { firstValueFrom } from 'rxjs'
import { toArray } from 'rxjs/operators'
import Adapter from './VcfTabixAdapter'
import configSchema from './configSchema'

test('adapter can fetch variants from volvox.vcf.gz', async () => {
  const adapter = new Adapter(
    configSchema.create({
      vcfGzLocation: {
        localPath: require.resolve('./test_data/volvox.filtered.vcf.gz'),
        locationType: 'LocalPathLocation',
      },
      index: {
        indexType: 'TBI',
        location: {
          localPath: require.resolve('./test_data/volvox.filtered.vcf.gz.tbi'),
          locationType: 'LocalPathLocation',
        },
      },
    }),
  )

  const csiAdapter = new Adapter(
    configSchema.create({
      vcfGzLocation: {
        localPath: require.resolve('./test_data/volvox.filtered.vcf.gz'),
        locationType: 'LocalPathLocation',
      },
      index: {
        indexType: 'CSI',
        location: {
          localPath: require.resolve('./test_data/volvox.filtered.vcf.gz.csi'),
          locationType: 'LocalPathLocation',
        },
      },
    }),
  )

  const csiFeatures = csiAdapter.getFeatures({
    refName: 'ctgA',
    start: 0,
    end: 20000,
  })

  const names = await adapter.getRefNames()
  const csiNames = await csiAdapter.getRefNames()
  expect(names).toEqual(csiNames)
  expect(names).toMatchSnapshot()

  const feat = adapter.getFeatures({
    refName: 'ctgA',
    start: 0,
    end: 20000,
  })

  const featArray = await firstValueFrom(feat.pipe(toArray()))
  const csiFeaturesArray = await firstValueFrom(csiFeatures.pipe(toArray()))
  expect(featArray.slice(0, 5)).toMatchSnapshot()
  expect(csiFeaturesArray.slice(0, 5)).toEqual(featArray.slice(0, 5))

  const featNonExist = adapter.getFeatures({
    refName: 'ctgC',
    start: 0,
    end: 20000,
  })

  const featArrayNonExist = await firstValueFrom(featNonExist.pipe(toArray()))
  expect(featArrayNonExist).toEqual([])
})
