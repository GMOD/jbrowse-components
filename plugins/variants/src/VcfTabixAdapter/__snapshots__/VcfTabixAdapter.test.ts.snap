// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch variants from volvox.vcf.gz 1`] = `
Array [
  "ctgA",
]
`;

exports[`adapter can fetch variants from volvox.vcf.gz 2`] = `
Array [
  Object {
    "ALT": Array [
      "C",
    ],
    "CHROM": "ctgA",
    "FILTER": null,
    "ID": null,
    "INFO": Object {
      "AC1": Array [
        2,
      ],
      "AF1": Array [
        1,
      ],
      "DP": Array [
        3,
      ],
      "DP4": Array [
        0,
        0,
        3,
        0,
      ],
      "FQ": Array [
        -36,
      ],
      "MQ": Array [
        37,
      ],
      "VDB": Array [
        0.0186,
      ],
    },
    "POS": 277,
    "QUAL": 10.4,
    "REF": "T",
    "aliases": undefined,
    "description": "SNV T -> C",
    "end": 277,
    "name": undefined,
    "refName": "ctgA",
    "samples": Object {
      "sample_data/raw/volvox/volvox-sorted.bam": Object {
        "GQ": Array [
          13,
        ],
        "GT": Array [
          "1/1",
        ],
        "PL": Array [
          42,
          9,
          0,
        ],
      },
    },
    "start": 276,
    "type": "SNV",
    "uniqueId": "test-vcf-604452",
  },
  Object {
    "ALT": Array [
      "C",
    ],
    "CHROM": "ctgA",
    "FILTER": null,
    "ID": null,
    "INFO": Object {
      "AC1": Array [
        2,
      ],
      "AF1": Array [
        1,
      ],
      "DP": Array [
        23,
      ],
      "DP4": Array [
        0,
        1,
        8,
        14,
      ],
      "FQ": Array [
        -73,
      ],
      "MQ": Array [
        37,
      ],
      "PV4": Array [
        1,
        1,
        0.42,
        1,
      ],
      "VDB": Array [
        0.0308,
      ],
    },
    "POS": 1694,
    "QUAL": 107,
    "REF": "T",
    "aliases": undefined,
    "description": "SNV T -> C",
    "end": 1694,
    "name": undefined,
    "refName": "ctgA",
    "samples": Object {
      "sample_data/raw/volvox/volvox-sorted.bam": Object {
        "GQ": Array [
          89,
        ],
        "GT": Array [
          "1/1",
        ],
        "PL": Array [
          140,
          46,
          0,
        ],
      },
    },
    "start": 1693,
    "type": "SNV",
    "uniqueId": "test-vcf-604550",
  },
  Object {
    "ALT": Array [
      "G",
    ],
    "CHROM": "ctgA",
    "FILTER": null,
    "ID": null,
    "INFO": Object {
      "AC1": Array [
        1,
      ],
      "AF1": Array [
        0.5,
      ],
      "DP": Array [
        22,
      ],
      "DP4": Array [
        6,
        5,
        3,
        8,
      ],
      "FQ": Array [
        45.3,
      ],
      "MQ": Array [
        37,
      ],
      "PV4": Array [
        0.39,
        1,
        1,
        1,
      ],
      "VDB": Array [
        0.0253,
      ],
    },
    "POS": 2644,
    "QUAL": 44,
    "REF": "T",
    "aliases": undefined,
    "description": "SNV T -> G",
    "end": 2644,
    "name": undefined,
    "refName": "ctgA",
    "samples": Object {
      "sample_data/raw/volvox/volvox-sorted.bam": Object {
        "GQ": Array [
          75,
        ],
        "GT": Array [
          "0/1",
        ],
        "PL": Array [
          74,
          0,
          77,
        ],
      },
    },
    "start": 2643,
    "type": "SNV",
    "uniqueId": "test-vcf-604667",
  },
  Object {
    "ALT": Array [
      "A",
    ],
    "CHROM": "ctgA",
    "FILTER": null,
    "ID": null,
    "INFO": Object {
      "AC1": Array [
        2,
      ],
      "AF1": Array [
        1,
      ],
      "DP": Array [
        19,
      ],
      "DP4": Array [
        0,
        0,
        8,
        11,
      ],
      "FQ": Array [
        -84,
      ],
      "MQ": Array [
        36,
      ],
      "VDB": Array [
        0.0384,
      ],
    },
    "POS": 3213,
    "QUAL": 124,
    "REF": "T",
    "aliases": undefined,
    "description": "SNV T -> A",
    "end": 3213,
    "name": undefined,
    "refName": "ctgA",
    "samples": Object {
      "sample_data/raw/volvox/volvox-sorted.bam": Object {
        "GQ": Array [
          99,
        ],
        "GT": Array [
          "1/1",
        ],
        "PL": Array [
          157,
          57,
          0,
        ],
      },
    },
    "start": 3212,
    "type": "SNV",
    "uniqueId": "test-vcf-604784",
  },
  Object {
    "ALT": Array [
      "ct",
    ],
    "CHROM": "ctgA",
    "FILTER": null,
    "ID": null,
    "INFO": Object {
      "AC1": Array [
        2,
      ],
      "AF1": Array [
        1,
      ],
      "DP": Array [
        15,
      ],
      "DP4": Array [
        0,
        0,
        5,
        6,
      ],
      "FQ": Array [
        -67.5,
      ],
      "INDEL": true,
      "MQ": Array [
        37,
      ],
      "VDB": Array [
        0.0384,
      ],
    },
    "POS": 3858,
    "QUAL": 160,
    "REF": "ctt",
    "aliases": undefined,
    "description": "deletion ctt -> ct",
    "end": 3860,
    "name": undefined,
    "refName": "ctgA",
    "samples": Object {
      "sample_data/raw/volvox/volvox-sorted.bam": Object {
        "GQ": Array [
          63,
        ],
        "GT": Array [
          "1/1",
        ],
        "PL": Array [
          201,
          33,
          0,
        ],
      },
    },
    "start": 3857,
    "type": "deletion",
    "uniqueId": "test-vcf-604886",
  },
]
`;
